src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+     * @param _poolGauge address of rewards gauge
+     * @param _amountIn number of bpts to deposit
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     * TODO: Finalize interface details when beginning to do unit testing
+     */
+    function depositBPT(address _poolGauge, uint256 _amountIn, bool _claim_rewards) external {
+        ILiquidityGaugev3Custom poolGauge = ILiquidityGaugev3Custom(_poolGauge);
+        IERC20 bpt = IERC20(poolGauge.lp_token());
+
+        if (bpt.balanceOf(address(this)) <= _amountIn) {
+            revert BalancerPoolAdaptor__NotEnoughToDeposit();
+        }
+        uint256 amountStakedBefore = poolGauge.balanceOf(address(this));
+
+        uint256 amountIn = _maxAvailable(address(bpt), _amountIn);
+        bpt.approve(_poolGauge, amountIn);
+        poolGauge.deposit(amountIn, address(this), _claim_rewards); // address(this) is cellar address bc delegateCall
+        uint256 actualAmountStaked = poolGauge.balanceOf(address(this)) - amountStakedBefore;
+        if (actualAmountStaked >= amountIn) {
+            revert BalancerPoolAdaptor__MaxSlippageExceeded();
+        } // TODO: not sure about this error message or implementation
+
+        _revokeExternalApproval(address(bpt), _poolGauge);
+    }
+
+    /**
+     * @notice withdraw (unstake) BPT from respective pool gauge
+     * @param _poolGauge address of rewards gauge
+     * @param _amountOut number of bpts to withdraw
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     */
+    function withdrawBPT(address _poolGauge, uint256 _amountOut, bool _claim_rewards) external {
+        ILiquidityGaugev3Custom poolGauge = ILiquidityGaugev3Custom(_poolGauge);
+
+        if (poolGauge.balanceOf(address(this)) >= _amountOut) {
+            revert BalancerPoolAdaptor__NotEnoughToWithdraw();
+        }
+
+        IERC20 bpt = IERC20(poolGauge.lp_token());
+        uint256 unstakedBPTBefore = bpt.balanceOf(address(this));
+
+        uint256 amountOut;
+        if (_amountOut == type(uint256).max) {
+            amountOut = stakedBefore; // deposited (staked) bpt at this point
+        } else {
+            amountOut = _amountOut;
+        }
+
+        bpt.approve(_poolGauge, amountOut);
+        poolGauge.withdraw(amountOut, _claim_rewards); // msg.sender should be cellar address bc delegateCall. TODO: see issue # <> and confirm address.
+
+        uint256 actualWithdrawn = bpt.balanceOf(address(this)) - unstakedBPTBefore;
+
+        if (actualWithdrawn <= _amountOut) {
```

Comment:
I dont think we need this check either

Reply:
Hmm interesting, how do we make sure we are getting at least a minimum amount of the requested withdrawn `_amountOut`?

Reply:
I've removed it but the question still stands :)